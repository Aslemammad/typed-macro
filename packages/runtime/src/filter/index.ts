export { FilterOptions, FilterPattern, createFilter, Filter } from './filter'
