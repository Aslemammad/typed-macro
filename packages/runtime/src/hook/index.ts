export { createHookContainer, HookContainer, CombinedHooks } from './hook'
