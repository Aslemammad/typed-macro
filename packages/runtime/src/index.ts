export * from './transformer'
export * from './normalizer'
export * from './renderer'
export * from './runtime'
export * from './loader'
export * from './filter'
export * from './hook'
