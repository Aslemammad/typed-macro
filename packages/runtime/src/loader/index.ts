export { Loader, createLoader } from './loader'
