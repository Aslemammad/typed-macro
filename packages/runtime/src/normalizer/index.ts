export {
  ModularizedMacros,
  ModularizedModules,
  ModularizedTypes,
  NormalizedExports,
  normalizeExports,
  normalizeMacroProvider,
  NormalizedMacroProvider,
} from './normalizer'
