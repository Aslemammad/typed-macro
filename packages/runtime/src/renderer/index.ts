export { TypeRenderer, createTypeRenderer } from './renderer'
