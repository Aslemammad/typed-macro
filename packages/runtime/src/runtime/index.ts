export { Runtime, RuntimeOptions, createRuntime } from './runtime'
