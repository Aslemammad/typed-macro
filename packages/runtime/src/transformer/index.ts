export {
  Transformer,
  createTransformer,
  TransformerOptions,
} from './transformer'
