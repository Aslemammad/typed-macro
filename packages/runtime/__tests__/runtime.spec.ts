import { createRuntime, Runtime } from '../src'
import { defineMacroProvider, EnvContext } from '@typed-macro/core'
import { createMacroRemove } from '../../core/__tests__/macros'

describe('Runtime', () => {
  const env: EnvContext = {
    host: 'test',
    packageManager: 'test',
    projectPath: [''],
    dev: true,
    ssr: false,
  }
  let runtime: Runtime
  beforeEach(() => {
    runtime = createRuntime(env, {})
  })

  it('should support add providers', () => {
    const provider = defineMacroProvider({
      id: 'provider',
      exports: {},
    })
    expect(() => runtime.appendProvider(provider)).not.toThrowError()
  })

  it('should support reject invalid providers', () => {
    expect(() => runtime.appendProvider({} as any)).toThrowError()
    const provider = defineMacroProvider({
      id: 'provider',
      exports: {},
    })
    provider.version = 'unknown'
    expect(() => runtime.appendProvider(provider)).toThrowError()
  })

  it('should support reject duplicated macros', () => {
    const provider = defineMacroProvider({
      id: 'provider',
      exports: {
        a: {
          macros: [createMacroRemove()],
        },
      },
    })
    runtime.appendProvider(provider)
    expect(() => runtime.appendProvider(provider)).toThrowError()
  })

  it('should support reject duplicated codes', () => {
    const provider = defineMacroProvider({
      id: 'provider',
      exports: {
        a: {
          code: `export let a = 1`,
        },
      },
    })
    runtime.appendProvider(provider)
    expect(() => runtime.appendProvider(provider)).toThrowError()
  })

  it('should support reject duplicated codes', () => {
    const provider = defineMacroProvider({
      id: 'provider',
      exports: {
        a: {
          code: `export let a = 1`,
        },
      },
    })
    runtime.appendProvider(provider)
    expect(() => runtime.appendProvider(provider)).toThrowError()
  })

  it('should work', async () => {
    const executedHooks: string[] = []
    const provider = defineMacroProvider({
      id: 'provider',
      exports: {
        a: { code: `export let a = 1` },
        b: { macros: [createMacroRemove()] },
      },
      hooks: {
        onStart() {
          executedHooks.push('hook.onStart')
        },
        onFilter(id) {
          executedHooks.push(`hook.onFilter: ${id}`)
          return true
        },
        beforeTransform(code, id) {
          executedHooks.push(`hook.beforeTransform: ${id}`)
          return code + '\nexport let a = 1'
        },
        afterTransform(code, id) {
          executedHooks.push(`hook.afterTransform: ${id}`)
          return code + '\nexport let b = 2'
        },
        onStop() {
          executedHooks.push('hook.onStop')
        },
      },
      options: {
        parserPlugins: ['decimal'],
      },
    })
    runtime.appendProvider(provider)

    await runtime.start()
    expect(runtime.resolveId('test.ts')).toBeUndefined()
    expect(runtime.resolveId('a')).toBe('a')
    await runtime.filter('test.ts')
    // filter reject .vue by default, so `onFilter` will be called
    await runtime.filter('test.vue')
    expect(runtime.load('a')).toMatchSnapshot()
    expect(
      await runtime.transform(
        `
    import { noop } from 'b'
    noop(0.3m)`,
        'test.ts'
      )
    ).toMatchSnapshot()
    await runtime.stop()
    expect(executedHooks).toMatchSnapshot()
  })
})
