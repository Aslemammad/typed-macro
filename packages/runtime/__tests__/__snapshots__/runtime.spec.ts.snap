// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Runtime should work 1`] = `
"export let a = 1
"
`;

exports[`Runtime should work 2`] = `
"


export let a = 1;
export let b = 2"
`;

exports[`Runtime should work 3`] = `
Array [
  "hook.onStart",
  "hook.onFilter: test.vue",
  "hook.beforeTransform: test.ts",
  "hook.afterTransform: test.ts",
  "hook.onStop",
]
`;
